_posts/2023-02-28-announcing-ballerina-2201.4.0-swan-lake-update-4.md

```diff
@@ -0,0 +1,109 @@
+
+```ballerina
+import ballerina/http;
+import ballerinax/azure_functions as af;
+
+
+type Album record {|
+    string id;
+    string title;
+    string artist;
+    decimal price;
+|};
+
+
+service / on new af:HttpListener() {
+    resource function post queue(@http:Payload Album album) returns [@af:HttpOutput http:Created, @af:QueueOutput {queueName: "albums"} string] {
+        http:Created httpRes = {
+            body: album.id + " Added to the queue!"
+        };
+        return [httpRes, album.id];
+    }
+}
+```
+
+This release comes with several bug fixes that address specification deviations present in previous versions. You may experience code breakage due to these fixes. Despite the potential for temporary disruption, it's critical to address these deviations to ensure the long-term stability and health of the language. 
+
+In previous versions, the compiler incorrectly identified the type of the following query action as `error?`.
+
+```ballerina
+public function main() {
+    error? e = from int i in 1 ... 10
+        do {
+            io:println(i);
+        };
+}
+```
+
+You can write the above code as follows with Update 4. 
+
+```ballerina
+public function main() {
+    from int i in 1 ... 3
+    do {
+        io:println(i);
+    };
+}
+```
+
+Here is another bug fix. Note the usage of the `check` keyword before the `from` keyword. In previous versions, the compiler includes the `error` type in the query action type when errors return from the `do` clause. 
+
+```ballerina
+function appendPRs(sheets:Client gsheets, PR[] pullRequests) returns error? {
+    error? e = from var {url, title, state, created_at, updated_at} in pullRequests
+        do {
+            check gsheets->appendRowToSheet(spreadSheetId, sheetName,
+                [url, title, state, created_at, updated_at]);
+        };
+    return e;
+}
+```
+
+You can write the above code as follows with Update 4. 
+
+```ballerina
+function appendPRs(sheets:Client gsheets, PR[] pullRequests) returns error? {
+    from var {url, title, state, created_at, updated_at} in pullRequests
+    do {
+        check gsheets->appendRowToSheet(spreadSheetId, sheetName,
+                [url, title, state, created_at, updated_at]);
+    };
+}
+```
+
+We added several new features to the standard library, including an experimental new feature that simplifies data persistence in Ballerina.  We believe this feature can potentially enhance your overall development experience. The concept is to declare the domain data model using Ballerina record types that serves as the definitive source of truth. This record type definition generates a client API to access and modify the data stored in MySQL, the currently supported data store. We will publish more information about this feature in a future blog post. See the [release note](https://ballerina.io/downloads/swan-lake-release-notes/swan-lake-2201.4.0) for more details of other standard library features. 
```

Comment:
Shouldn't we link https://ballerina.io/downloads/swan-lake-release-notes/swan-lake-2201.4.0#standard-library-updates instead of the above link?

Reply:
Updated.